documentation/Using-the-SDK/Advanced/Creating-Your-Own-DataSource.md

```diff
@@ -0,0 +1,216 @@
+# Abstract
+
+This document outlines a scenario where you want to implement a `DataSource`
+beyond the `FileDataSource` and `DirectoryDataSource` provided by the SDK.
+
+# Motivation
+
+For users that have data stored somewhere other than files or directories, it is
+desirable for the SDK to treat their `DataSource` as a first class `DataSource`.
+For example, suppose that you are writing many `CustomDataSource`s that get
+their data from a SQL database or from a web service. Rather than having a file
+with a URI in it, it would be nice to say that your `Custom DataSource` uses a 
+`SqlServerDataSource` or a `WebServiceDataSource`.
+
+# Requirements
+
+In order to have your new `DataSource` recognized by the SDK, you must do the
+following:
+
+- Create a new class inheriting from `DataSource`.
+    - You may also implment `IDataSource` directly.
+- Create a new Attribute inheriting from `DataSourceAttribute`, and specify the
+  type of the class you created in the call to the base constructor.
+- Optionally implement the `Accepts` method on your new attribute.
+
+Once you have implemented these three items, the SDK will recognize your new
+`DataSource` and route it to `CustomDataSource`s as appropriate.
+
+# Implementation
+
+This section outlines implementing a new `DataSource` type to be
+consumed by `CustomDataSource`s. For the purposes of this walkthrough, we will
+be implementing a `DataSource` backed by data from a web service. This example will
+- Create a new `WebServiceDataSource` class derived from `DataSource`.
+- Create a new `WebServiceDataSourceAttribute` so that `CustomDataSource`s
+  can consume `WebServiceDataSource`s
+- Implement `Accepts` on the `WebServiceDataSourceAttribute`
+
+## DataSource class
+
+We first create our `WebServiceDataSource` class. This is how users tell us where the
+Web Service is located.
```

Comment:
1) lowercase "Web Service"
2) Maybe say "where the web service providing data is located"